builtin/async/common.lua

```diff
@@ -0,0 +1,18 @@
+-- NOTE: This script is used by the main thread and async worker threads
+
+-- Registered metatables, used in C++ for deserialization
+core.registered_async_metatables = {}
+function core.register_async_metatable(name, mt)
+	if type(name) ~= "string" then
+		return error(("attempt to use %s value as metatable name"):format(type(mt)))
+	elseif type(mt) ~= "table" then
+		return error(("attempt to register a %s value as metatable"):format(type(mt)))
+	elseif core.registered_async_metatables[name] ~= nil then
+		return error(("attempt to override metatable %s"):format(name))
+	elseif core.registered_async_metatables[mt] ~= nil then
+		return error("attempt to register metatable twice")
```

Comment:
idea: you could make it valid to register a mt twice as long as the values are equal

Reply:
Like, allowing a single metatable to have different names? I agree that it would make sense in certain usecases.

Reply:
I mean make it valid to do the following:
```lua
core.register_async_metatable("__builtin:vector", vector.metatable)
core.register_async_metatable("__builtin:vector", vector.metatable)
```
but not this
```lua
core.register_async_metatable("__builtin:vector", vector.metatable)
core.register_async_metatable("__builtin:vector2", vector.metatable)
-- or
core.register_async_metatable("__builtin:vector", vector.metatable)
core.register_async_metatable("__builtin:vector", {funny = "table"})
```

Reply:
Ok, turns out I got this reply when I was already working on this.

For 4fc668b7d5b65766b5bc6a81ac8fe5bb27cf855e, I chose to allow registering a metatable multiple times, including under different names. The latter seems useful in e.g. situations where a mod depending on an API library registers the metatable of the library to not wait wait for it to be updated/released. That said, something like this should ideally be avoided.

The same commit includes the test case for the above situation as well as for not allowing the latter situation you mentioned.